### PR TITLE
Change to portable network proto using a serializer

### PR DESCRIPTION
**Pull request by ganwell**
This commit is missing the following:

* Unittests

I request the following review:

* Please check if my pointer arithmetic style is portable, since
  paranoid C doesn't allow "void* a = x; a += 3;" we have to move
  through the buffer in a "char" context and then cast to void*

* Please do a network research to see if you come to the same conclusion
  as I.

* Please review the actual code (all: serializer.rg.h, sosterwalder: everything)

About the problem:

It seems that structs are mostly portable, but not in the least guaranteed
to be portable.
https://stackoverflow.com/questions/45116212/are-packed-structs-portable

What is currently done:

* Assume the layout and size of struct is the same on the target
  platforms

* Change every field that goes to the network to network-order and vice
  versa

What should be done:

* Change every field that goes to the network to network-order and vice
  versa

* Layout and read the data manually, meaning write a manual
  (de-)serializer, that is guaranteed to do the same layout on every
  platform

* Don't use sizeof the message, but the size we learned by designing the
  serializer (a constant)

No, using msgpack or anything similar is not a good idea, since most
implementation are too complex and often not portable (not the
serialization format is not portable, but the implementation).

### CODE REVIEW

**Comment by ganwell** [comment]
@winged @keachi @sosterwalder I am quite sure I did the right thing, but opinions on the net about this are so inconsistent, that I need a review from all of you. Especially:

What we all used to do, add numbers to void* pointer, isn't allowed in C99. So I assume the correct thing to do is:

* Navigate the buffer in char* context (we know the size is 1 byte)
* Then get the pointer
* Assign the pointer to actual datatype

File: src/serializer.rg.h [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    CH_SR_WIRE_MESSAGE_CHECK
+#enddef
+
+// Definitions
+// ===========
+
+// .. c:function::
+static
+ch_inline
+void
```
Comment (winged) [inline src/serializer.rg.h]: I would return the number of bytes read, so your caller doesn't need to calculate the offset where to continue reading itself.

File: src/serializer.rg.h [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+    msg->type = *type;
+    msg->header_len = ntohs(*header_len);
+    msg->data_len = ntohl(*data_len);
+}
+
+// .. c:function::
+static
+ch_inline
+void
```
Comment (winged) [inline src/serializer.rg.h]: Same as above - return the number of bytes read

**Comment by ganwell** [comment]
Discussed offline with the people.

**Pull request merged**